sql/insert.go

```diff
@@ -60,32 +62,52 @@ func (p *planner) Insert(n *parser.Insert) (planNode, error) {
 		return nil, err
 	}
 
+	primaryIndex := tableDesc.Indexes[0]
+	primaryIndexKeyPrefix := encodeIndexKeyPrefix(tableDesc.ID, primaryIndex.ID)
+
 	b := client.Batch{}
+
 	for rows.Next() {
 		values := rows.Values()
 		if len(values) != len(cols) {
 			return nil, fmt.Errorf("invalid values for columns: %d != %d", len(values), len(cols))
 		}
-		indexKey := encodeIndexKeyPrefix(desc.ID, desc.Indexes[0].ID)
-		primaryKey, err := encodeIndexKey(desc.Indexes[0], colMap, values, indexKey)
+
+		primaryIndexKeySuffix, err := encodeIndexKey(primaryIndex, colMap, values, nil)
+		if err != nil {
+			return nil, err
+		}
+		primaryIndexKey := bytes.Join([][]byte{primaryIndexKeyPrefix, primaryIndexKeySuffix}, nil)
+
+		// Write the secondary indexes.
+		secondaryIndexEntries, err := encodeSecondaryIndexes(tableDesc, colMap, values, primaryIndexKeySuffix)
 		if err != nil {
 			return nil, err
 		}
+
+		for _, secondaryIndexEntry := range secondaryIndexEntries {
+			if log.V(2) {
+				log.Infof("CPut %q -> %v", secondaryIndexEntry.key, secondaryIndexEntry.value)
+			}
+			b.CPut(secondaryIndexEntry.key, secondaryIndexEntry.value, nil)
+		}
+
+		// Write the row.
 		for i, val := range values {
-			key := encodeColumnKey(cols[i], primaryKey)
+			key := encodeColumnKey(cols[i], primaryIndexKey)
 			if log.V(2) {
-				log.Infof("Put %q -> %v", key, val)
+				log.Infof("CPut %q -> %v", key, val)
 			}
 			// TODO(pmattis): Need to convert the value type to the column type.
 			switch t := val.(type) {
 			case parser.DBool:
-				b.Put(key, bool(t))
+				b.CPut(key, bool(t), nil)
 			case parser.DInt:
-				b.Put(key, int64(t))
+				b.CPut(key, int64(t), nil)
 			case parser.DFloat:
-				b.Put(key, float64(t))
+				b.CPut(key, float64(t), nil)
 			case parser.DString:
-				b.Put(key, string(t))
+				b.CPut(key, string(t), nil)
```

Comment:
I thought the primary index is by definition `Indexes[0]`.

On Mon, Aug 3, 2015 at 1:20 PM, vivekmenezes <notifications@github.com>
wrote:

> In sql/insert.go
> <https://github.com/cockroachdb/cockroach/pull/1886#discussion_r36107777>:
>
> >  			case parser.DString:
> > -				b.Put(key, string(t))
> > +				b.CPut(key, string(t), nil)
>
> Hey Pete, I happen to do it and came up with some puzzling results.
> Actually what i did was incorrect, but turns out I've uncovered a bug with
> us assuming that index[0] is always a primary index. I'm working on a PR to
> get that fixed. Vivek
> … <#14ef4948501528d9_>
> On Mon, Aug 3, 2015 at 12:56 PM Peter Mattis <notifications@github.com>
> wrote: In sql/insert.go <
> https://github.com/cockroachdb/cockroach/pull/1886#discussion_r36105364>:
> > case parser.DString: > - b.Put(key, string(t)) > + b.CPut(key, string(t),
> nil) Why would changing these CPuts to Puts be valid? We need this to be
> CPuts so that insertion of a duplicate row is detected. Am I missing
> something? — Reply to this email directly or view it on GitHub <
> https://github.com/cockroachdb/cockroach/pull/1886/files#r36105364>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/1886/files#r36107777>.
>